src/runtime/task/mod.rs

```diff
@@ -175,31 +176,35 @@ impl From<TaskId> for usize {
 // TODO this probably won't work well with large numbers of tasks -- maybe a BitVec?
 #[derive(PartialEq, Eq)]
 pub(crate) struct TaskSet {
-    tasks: SmallVec<[bool; MAX_INLINE_TASKS]>,
+    tasks: BitVec,
 }
 
 impl TaskSet {
-    pub fn new(max_tasks: usize) -> Self {
+    pub fn new() -> Self {
         Self {
-            // Need to create the inner vec! to ensure elements are initialized
-            tasks: SmallVec::from_vec(vec![false; max_tasks]),
+            tasks: BitVec::from_bitslice(bits![0; DEFAULT_INLINE_TASKS]),
         }
     }
 
     pub fn contains(&self, tid: TaskId) -> bool {
-        self.tasks[tid.0]
+        // Return false if tid is outside the TaskSet
+        (tid.0 < self.tasks.len()) && self.tasks[tid.0]
     }
 
     pub fn is_empty(&self) -> bool {
         self.tasks.iter().all(|b| !*b)
     }
 
     pub fn insert(&mut self, tid: TaskId) {
-        self.tasks[tid.0] = true;
+        if tid.0 >= self.tasks.len() {
+            self.tasks.resize(2 * self.tasks.len(), false);
+        }
+        assert!(self.tasks.len() > tid.0);
```

Comment:
What happens if our current size is 32 and then we ask to insert task 70? I think we should just resize to length `tid.0 + 1` and let the `BitVec` decide how it wants to amortize that cost (it just delegates to `Vec` IIRC).

Reply:
Good catch; fixing